src/pybind/mgr/pg_autoscaler/module.py

```diff
@@ -0,0 +1,468 @@
+                self.osd_count = None  # Number of OSDs
+                self.pg_target = None  # Ideal full-capacity PG count?
+                self.pg_current = 0  # How many PGs already?
+                self.capacity = None  # Total capacity of OSDs in subtree
+                self.pool_ids = []
+                self.pool_names = []
+
+        # identify subtrees (note that they may overlap!)
+        for pool_id, pool in osdmap.get_pools().items():
+            cr_name = crush.get_rule_by_id(pool['crush_rule'])['rule_name']
+            root_id = int(crush.get_rule_root(cr_name))
+            pool_root[pool_id] = root_id
+            osds = set(crush.get_osds_under(root_id))
+
+            # do we intersect an existing root?
+            s = None
+            for prev in result.itervalues():
+                if osds & prev.osds:
+                    s = prev
+                    break
+            if not s:
+                s = CrushSubtreeResourceStatus()
+                roots.append(s)
+            result[root_id] = s
+            s.root_ids.append(root_id)
+            s.osds |= osds
+            s.pool_ids.append(int(pool_id))
+            s.pool_names.append(pool['pool_name'])
+            s.pg_current += pool['pg_num_target'] * pool['size'];
+
+        # FIXME: avoid dumping out the whole PG map to python objects, use df
+        # instead
+        pg_map = self.get('pg_dump')
+
+        # finish subtrees
+        for s in roots:
+            s.osd_count = len(s.osds)
+            s.pg_target = s.osd_count * int(self.mon_target_pg_per_osd)
+
+            capacity = 0.0
+            for osd_stats in pg_map['osd_stats']:
+                if osd_stats['osd'] in s.osds:
+                    # Intentionally do not apply the OSD's reweight to
+                    # this, because we want to calculate PG counts based
+                    # on the physical storage available, not how it is
+                    # reweighted right now.
+                    capacity += osd_stats['kb'] * 1024
+
+            s.capacity = capacity
+
+            self.log.debug('root_ids %s pools %s with %d osds, pg_target %d',
+                           s.root_ids,
+                           s.pool_ids,
+                           s.osd_count,
+                           s.pg_target)
+
+        return result, pool_root
+
+
+    def _get_pool_status(
+            self,
+            osdmap,
+            pools,
+            threshold=3.0,
+    ):
+        assert threshold >= 2.0
+
+        crush_map = osdmap.get_crush()
+
+        root_map, pool_root = self.get_subtree_resource_status(osdmap, crush_map)
+
+        df = self.get('df')
+        pool_stats = dict([(p['id'], p['stats']) for p in df['pools']])
+
+        ret = []
+
+        # iterate over all pools to determine how they should be sized
+        for pool_name, p in pools.iteritems():
+            pool_id = p['pool']
+
+            # FIXME: we assume there is only one take per pool, but that
```

Comment:
is there any common crush rule like that which should be mentioned in the docs?

Reply:
not really...